pkg/operator/controller/extension/virtualcluster/add.go

```diff
@@ -0,0 +1,98 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package virtualcluster
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/clock"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	operatorv1alpha1 "github.com/gardener/gardener/pkg/apis/operator/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	clientmapbuilder "github.com/gardener/gardener/pkg/client/kubernetes/clientmap/builder"
+	"github.com/gardener/gardener/pkg/controllerutils/mapper"
+	operatorpredicate "github.com/gardener/gardener/pkg/operator/predicate"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "garden-extension-virtual-cluster"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager) error {
+	if r.RuntimeClient == nil {
+		runtimeClientSet, err := kubernetes.NewWithConfig(
+			kubernetes.WithRESTConfig(mgr.GetConfig()),
+			kubernetes.WithRuntimeAPIReader(mgr.GetAPIReader()),
+			kubernetes.WithRuntimeClient(mgr.GetClient()),
+			kubernetes.WithRuntimeCache(mgr.GetCache()),
+		)
+		if err != nil {
+			return fmt.Errorf("failed creating runtime clientset: %w", err)
+		}
+		r.RuntimeClient = runtimeClientSet.Client()
+	}
+
+	if r.Clock == nil {
+		r.Clock = clock.RealClock{}
+	}
+	if r.Recorder == nil {
+		r.Recorder = mgr.GetEventRecorderFor(ControllerName + "-controller")
+	}
+	if r.GardenNamespace == "" {
+		r.GardenNamespace = v1beta1constants.GardenNamespace
+	}
+	if r.GardenClientMap == nil {
+		var err error
+		r.GardenClientMap, err = clientmapbuilder.
+			NewGardenClientMapBuilder().
+			WithRuntimeClient(r.RuntimeClient).
+			WithClientConnectionConfig(&r.Config.VirtualClientConnection).
+			WithGardenNamespace(r.GardenNamespace).
+			Build(mgr.GetLogger())
+		if err != nil {
+			return fmt.Errorf("failed to build garden ClientMap: %w", err)
+		}
```

Comment:
Can you rather make this client map an argument in the `AddToManager` func? I think we follow this pattern in `gardenlet` when it's about the shoot clientmap.

Reply:
The format here matches what the other controllers in the operator use e.g. [care](https://github.com/gardener/gardener/blob/d5e74137890a7144d7429ed6c9a5e289191610d5/pkg/operator/controller/garden/care/add.go#L38)

Reply:
> to take gardenClientMap as argument to AddToManager

I did a small overall refactoring (like the creation on `gardenClientMap` in `app.go`), but did not change the `AddToManager` because I think it does not fit. Overall, even in the gardenlet code I see the same pattern:
- you inject the `gardenClientMap` in the "intermediate" package level `AddToManager` funcs, but your reconcilers just get them injected directly in the struct e.g. https://github.com/gardener/gardener/blob/93bd7a66d019c61eecdbc8bba4b9ce560da40b9b/pkg/gardenlet/controller/shoot/add.go#L63 and this is the same case. Intuitively it makes sense because I need to store the `shootClientMap` in the struct as I need it for the operations. 
 
For all intents and purposes, the initialisation in the commented snippet above can be omitted and return an error if nil though. But the reconciler basically tries to initialise all that was not injected.